src/actions/pause/mapScopes.js

```diff
@@ -24,57 +27,91 @@ import type {
 
 import type { ThunkArgs } from "../types";
 
-export function mapScopes(scopes: Scope, frame: Frame) {
+export function mapScopes(scopes: Promise<Scope>, frame: Frame) {
   return async function({ dispatch, getState, client, sourceMaps }: ThunkArgs) {
-    const sourceRecord = getSource(getState(), frame.location.sourceId);
-    await dispatch(loadSourceText(sourceRecord));
-    const source = sourceRecord.toJS();
-
-    const originalAstScopes = await getScopes(frame.location);
-    const generatedAstScopes = await getScopes(frame.generatedLocation);
-
-    if (!originalAstScopes || !generatedAstScopes) {
-      return;
-    }
-
-    const generatedAstBindings = buildGeneratedBindingList(
-      scopes,
-      generatedAstScopes
+    const generatedSourceRecord = getSource(
+      getState(),
+      frame.generatedLocation.sourceId
     );
 
-    const mappedOriginalScopes = await Promise.all(
-      Array.from(originalAstScopes, async item => {
-        const generatedBindings = {};
-
-        await Promise.all(
-          Object.keys(item.bindings).map(async name => {
-            generatedBindings[name] = await findGeneratedBinding(
-              sourceMaps,
-              source,
-              name,
-              item.bindings[name],
-              generatedAstBindings
-            );
-          })
-        );
-
-        return {
-          ...item,
-          generatedBindings
-        };
-      })
-    );
+    const sourceRecord = getSource(getState(), frame.location.sourceId);
 
-    const mappedScopes = generateClientScope(scopes, mappedOriginalScopes);
+    const shouldMapScopes =
+      features.mapScopes &&
+      !generatedSourceRecord.get("isWasm") &&
+      !sourceRecord.get("isPrettyPrinted") &&
+      !isGeneratedId(frame.location.sourceId);
 
     dispatch({
       type: "MAP_SCOPES",
       frame,
-      scopes: mappedScopes
+      [PROMISE]: (async function() {
+        let mappedScopes;
+        if (shouldMapScopes) {
```

Comment:
minor thing:

```js
if (!shouldMapScopes) {
  return scopes;
}

...
return mappedScopes
```

Reply:
I didn't do that at the time because the `|| scopes` logic is needed even when `shouldMapScopes` is true, but since we talked about changing that to store `null` and have the selector fallback to generated, I can definitely do `return null` early here now.

I generally find that early returns any time there's something more complicated than `null` just end up leading to cases where you assume something always runs and end up being wrong.

Reply:
lol that is fair